feat(hardware control): create logic to detect if tip has hit bottom …

aaron-kulkarni opened the pull request:
…of well or liquid

Creates a new file called sensor_utils.py. Introduces new error type for tip hitting bottom of the well when searching for liquid. Creates a function that detects whether the tip has hit a liquid or the bottom of the well based on pressure data.

fix EXEC-517

<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Test Plan

<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

aaron-kulkarni committed:
feat(hardware control): create logic to detect if tip has hit bottom of well or liquid

Creates a new file called sensor_utils.py. Introduces new error type for tip hitting bottom of the
well when searching for liquid. Creates a function that detects whether the tip has hit a liquid or
the bottom of the well based on pressure data.

fix EXEC-517

sfoster1 commented on hardware/opentrons_hardware/hardware_control/sensor_utils.py:
```diff
@@ -0,0 +1,27 @@
+import numpy as np
+
+
+def detectLiquidOrBlockage(
```
our style is `snake_case` for python, so this should be `detect_liquid_or_blockage`, same with the variables (sorry!)

sfoster1 commented on hardware/opentrons_hardware/hardware_control/sensor_utils.py:
```diff
@@ -0,0 +1,27 @@
+def detectLiquidOrBlockage(
+        pressureReadings: list[float],
+        liquidSolidThreshold: float
+) -> None:
```
I think we want this function to return a `bool`, and probably have the function be _named_ in a way that can be answered with true or false - i.e., `did_hit_well_bottom(...) -> bool` or something. 

The reason to make it return a bool is that a function that returns a bool is more generally useful. A function that either raises or returns can't be used for anything else without doing some exception-as-control-flow mechanics we don't like doing; but if the function returns a bool, the caller can raise if the bool is false

sfoster1 commented on hardware/opentrons_hardware/hardware_control/sensor_utils.py:
```diff
@@ -0,0 +1,27 @@
+        liquidSolidThreshold: float
+) -> None:
+    dx = 1
+    #take derivative of pressureReadings with respect to dx
```
This comment and the variables names talk about derivatives, but this is a discrete gradient and thus operates in terms of _difference_. It could be quite relevant, given that when you're living in discrete world with noise, differencing amplifies noise. Let's phrase things as differences to keep that in mind.

I also think this comment is duplicative with the next line; if you know what `dx` and a gradient are, you can understand the next line, and if you don't then this comment doesn't tell you anything new.

sfoster1 reviewed (changes requested):
Great job with the enumerated error! I think there's some things to fix up about the functions though.

sfoster1 commented:
Also, the thing with the commit titles is 
- they want to have a type that is correct (which this is)
- but the scope, i.e. what's inside the parens, actually has to be exactly the name of a subdirectory of the monorepo 

So here it would be exactly `feat(hardware):`

aaron-kulkarni committed:
refactor(hardware): change function definition and variable names for improved formatting

Refactor detectLiquidOrBlockage to did_tip_hit_liquid among other minor formatting changes

re EXEC-517

sfoster1 commented on hardware/opentrons_hardware/hardware_control/sensor_utils.py:
```diff
@@ -0,0 +1,25 @@
+        pressure_readings: list[float],
+        liquid_solid_threshold: float
+) -> bool:
+    pressure_difference = np.gradient(pressure_readings[-5:], 1)
```
What happens if there's fewer than 5 readings in the list?

ryanthecoder replied:
there should always be a minimum of 100 in the future implementation but a check here makes sense, maybe we can raise a LiquidNotFoundError if there isn't enough data

sfoster1 reviewed (changes requested):
I think we need to handle the length of the list in some way - either gracefully degrade under 5 samples, or raise an error

aaron-kulkarni committed:
refactor(hardware): add error for not enough pressure data

Add a LiquidNotFoundErorr when the pressure data passed into did_tip_hit_liquid is too short

re EXEC-517

sfoster1 reviewed (changes requested):
The code looks good to me, so now we just need to get the static analysis passing. You should autoformat the code with `make -C api format`, and see what `make -C api lint` gives you; same with `make -C shared-data/python format` and `make -C shared-data/python lint` (and a top level `make format-js` to get the errors.json file autoformatted)

aaron-kulkarni committed:
lint

aaron-kulkarni committed:
added docstring

aaron-kulkarni committed:
added another docstring

sfoster1 reviewed (approved):
Looks good to me! Nice work. Future stuff will require tests but here we're making exactly one call to numpy so I think it would be a little much :)
